src/lib/core/fiber.c

```diff
@@ -653,6 +653,15 @@ fiber_join_timeout(struct fiber *fiber, double timeout)
 		} while (!fiber_is_dead(fiber) && !exceeded);
 	}
 
+	if ((fiber->flags & FIBER_IS_JOINABLE) == 0) {
+		/*
+		 * Joined by concurrent fiber_join_timeout.
+		 * This should never happen when called from lbox_fiber_join.
+		 */
+		diag_set(IllegalParams, "the fiber is already joined");
+		return -1;
+	}
+
```

Comment:
The problem is the same here as it was in Lua - you can't access `fiber` if it was joined by somebody else already. It will crash or hang, depending on whether the fiber was already reused or just deleted. This slight alteration of the test makes it crash:
```Diff
-	struct fiber *joinee = fiber_new_xc("joinee", noop_f);
+	struct fiber_attr *attr = fiber_attr_new();
+	fail_if(fiber_attr_setstacksize(attr, 1024 * 1024) != 0);
+	struct fiber *joinee = fiber_new_ex("joinee", attr, noop_f);
+	fail_if(joinee == NULL);
+	fiber_attr_delete(attr);
```
It is far from trivial to fix for the public C API. Just doing `fiber_find()` after each yield doesn't look like a good solution in terms of perf. I tried to explain in the ticket our other options. If we don't want to wrap private struct fiber into a safe public version, then maybe we could reference-count the internal fibers. Or just leave the issue as is, I suppose. I won't insist.

This particular commit could be omitted, the problem remains anyway. If the fiber was reused and also became joinable, then you will join a totally foreign fiber. If the fiber was deleted - you will crash.

Reply:
I removed this patch and created #7562 about C API.